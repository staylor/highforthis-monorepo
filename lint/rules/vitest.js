import { OFF, ERROR } from './constants.js';

export default {
  'vitest/consistent-test-it': [
    ERROR,
    {
      fn: 'it',
      withinDescribe: 'it',
    },
  ],
  'vitest/max-expects': OFF,
  'vitest/no-conditional-expect': OFF,
  'vitest/no-conditional-in-test': OFF,
  'vitest/no-disabled-tests': OFF,
  'vitest/no-hooks': OFF,
  'vitest/prefer-called-with': OFF,
  'vitest/prefer-expect-assertions': OFF,
  'vitest/prefer-strict-equal': OFF,
  'vitest/require-hook': OFF,
  'vitest/prefer-to-be-falsy': OFF,
};
