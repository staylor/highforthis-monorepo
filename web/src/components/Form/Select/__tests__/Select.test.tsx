import { screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import Select from '@/components/Form/Select';

const placeholder = '-- Select Flavor --';
const flavors = ['Mild', 'Hot', 'Fire'];
const flavorMap = [
  { label: 'Mild', value: 'mild' },
  { label: 'Hot', value: 'hot' },
  { label: 'Fire', value: 'fire' },
];
const tacoMap = [
  { label: 'Crunchy', value: 'crunchy' },
  { label: 'Soft', value: 'soft' },
];
const groups = [
  { label: 'Tacos', choices: tacoMap },
  { label: 'Saunce', choices: flavorMap },
];

describe('select', () => {
  it('empty', () => {
    const { container } = render(<Select />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('add className', () => {
    // eslint-disable-next-line tailwindcss/no-custom-classname
    const { container } = render(<Select className="foo" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  describe('placeholder', () => {
    it('no choices', () => {
      const { container } = render(<Select placeholder={placeholder} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('choices', () => {
      const { container } = render(<Select placeholder={placeholder} choices={flavors} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('choices as objects', () => {
      const { container } = render(<Select placeholder={placeholder} choices={flavorMap} />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('multiple', () => {
    it('true', () => {
      const { container } = render(<Select multiple />);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('false', () => {
      const { container } = render(<Select multiple={false} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('string', () => {
      // @ts-ignore
      const { container } = render(<Select multiple="multiple" />);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('truthy', () => {
      // @ts-ignore
      const { container } = render(<Select multiple="1" />);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('falsey', () => {
      // @ts-ignore
      const { container } = render(<Select multiple="0" />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('onChange', () => {
    it('value', async () => {
      const user = userEvent.setup();
      const func = vi.fn();
      render(<Select onChange={func} choices={flavorMap} />);
      const value = 'fire';

      await user.selectOptions(screen.getByRole('combobox'), [value]);

      expect(func).toHaveBeenCalledWith(value);
    });

    it('values', async () => {
      const user = userEvent.setup();
      const selectSpy = {
        onChange: () => null,
      };

      const func = vi.spyOn(selectSpy, 'onChange');
      render(<Select multiple onChange={func as any} choices={flavorMap} />);

      await user.selectOptions(screen.getByRole('listbox'), ['mild', 'fire']);

      expect(func).toHaveBeenCalledWith(['mild', 'fire']);
    });
  });

  describe('choices', () => {
    it('choices', () => {
      const { container } = render(<Select choices={flavors} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('choices as objects', () => {
      const { container } = render(<Select choices={flavorMap} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('choice', () => {
      const value = 'medium';
      const { container } = render(<Select choices={flavorMap} value={value} />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('groups', () => {
    it('no value', () => {
      const { container } = render(<Select groups={groups} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('value', () => {
      const value = 'mild';
      const { container } = render(<Select groups={groups} value={value} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('values', () => {
      const value = ['soft', 'mild'];
      const { container } = render(<Select multiple groups={groups} value={value} />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('child nodes', () => {
    it('children', () => {
      const { container } = render(
        <Select>
          <option value="meximelt">Mexi-melt</option>
        </Select>
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('children with choices', () => {
      const { container } = render(
        <Select choices={flavorMap}>
          <option value="meximelt">Mexi-melt</option>
        </Select>
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
