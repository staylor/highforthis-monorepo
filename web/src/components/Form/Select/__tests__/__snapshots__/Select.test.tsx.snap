// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`select > add className 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors foo"
/>
`;

exports[`select > child nodes > children 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value="meximelt"
  >
    Mexi-melt
  </option>
</select>
`;

exports[`select > child nodes > children with choices 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value="mild"
  >
    Mild
  </option>
  <option
    value="hot"
  >
    Hot
  </option>
  <option
    value="fire"
  >
    Fire
  </option>
  <option
    value="meximelt"
  >
    Mexi-melt
  </option>
</select>
`;

exports[`select > choices > choice 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value="mild"
  >
    Mild
  </option>
  <option
    value="hot"
  >
    Hot
  </option>
  <option
    value="fire"
  >
    Fire
  </option>
</select>
`;

exports[`select > choices > choices 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value="Mild"
  >
    Mild
  </option>
  <option
    value="Hot"
  >
    Hot
  </option>
  <option
    value="Fire"
  >
    Fire
  </option>
</select>
`;

exports[`select > choices > choices as objects 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value="mild"
  >
    Mild
  </option>
  <option
    value="hot"
  >
    Hot
  </option>
  <option
    value="fire"
  >
    Fire
  </option>
</select>
`;

exports[`select > empty 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
/>
`;

exports[`select > groups > no value 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <optgroup
    label="Tacos"
  >
    <option
      value="crunchy"
    >
      Crunchy
    </option>
    <option
      value="soft"
    >
      Soft
    </option>
  </optgroup>
  <optgroup
    label="Saunce"
  >
    <option
      value="mild"
    >
      Mild
    </option>
    <option
      value="hot"
    >
      Hot
    </option>
    <option
      value="fire"
    >
      Fire
    </option>
  </optgroup>
</select>
`;

exports[`select > groups > value 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <optgroup
    label="Tacos"
  >
    <option
      value="crunchy"
    >
      Crunchy
    </option>
    <option
      value="soft"
    >
      Soft
    </option>
  </optgroup>
  <optgroup
    label="Saunce"
  >
    <option
      value="mild"
    >
      Mild
    </option>
    <option
      value="hot"
    >
      Hot
    </option>
    <option
      value="fire"
    >
      Fire
    </option>
  </optgroup>
</select>
`;

exports[`select > groups > values 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
  multiple=""
>
  <optgroup
    label="Tacos"
  >
    <option
      value="crunchy"
    >
      Crunchy
    </option>
    <option
      value="soft"
    >
      Soft
    </option>
  </optgroup>
  <optgroup
    label="Saunce"
  >
    <option
      value="mild"
    >
      Mild
    </option>
    <option
      value="hot"
    >
      Hot
    </option>
    <option
      value="fire"
    >
      Fire
    </option>
  </optgroup>
</select>
`;

exports[`select > multiple > false 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
/>
`;

exports[`select > multiple > falsey 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
  multiple=""
/>
`;

exports[`select > multiple > string 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
  multiple=""
/>
`;

exports[`select > multiple > true 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
  multiple=""
/>
`;

exports[`select > multiple > truthy 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
  multiple=""
/>
`;

exports[`select > placeholder > choices 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value=""
  >
    -- Select Flavor --
  </option>
  <option
    value="Mild"
  >
    Mild
  </option>
  <option
    value="Hot"
  >
    Hot
  </option>
  <option
    value="Fire"
  >
    Fire
  </option>
</select>
`;

exports[`select > placeholder > choices as objects 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value=""
  >
    -- Select Flavor --
  </option>
  <option
    value="mild"
  >
    Mild
  </option>
  <option
    value="hot"
  >
    Hot
  </option>
  <option
    value="fire"
  >
    Fire
  </option>
</select>
`;

exports[`select > placeholder > no choices 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value=""
  >
    -- Select Flavor --
  </option>
</select>
`;
